packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js

```diff
@@ -9,8 +9,13 @@ let buildRequire = template(`
 `);
 
 let buildExportsModuleDeclaration = template(`
-  Object.defineProperty(exports, "__esModule", {
-    value: true
+  Object.defineProperties(exports, {
+    __esModule: {
+      value: true
+    },
+    __esExports: {
+      value: $0
```

Comment:
You should be able to use:

    // At a point early on so it exists when you call `buildExportAll`
    var namedExportIdentifier = path.scope.generateUidIdentifier('namedExports');

    // At the bottom 
    path.scope.push({id: namedExportIdentifier, init: t.valueToNode(Object.keys(exportNames))});

That's from memory, so if you have issues let me know.